lib/wibox/layout/ratio.lua

```diff
@@ -0,0 +1,339 @@
+-- @name add
+-- @class function
+
+--- Fit the fixed layout into the given space
+-- @param layout The layout you are modifying.
+-- @param context The context in which we are fit.
+-- @tparam number orig_width The available width.
+-- @tparam number orig_height The available height.
+-- @name fit
+-- @class function
+
+--- Reset a fixed layout. This removes all widgets from the layout.
+-- @param layout The layout you are modifying.
+-- @name reset
+-- @class function
+
+-- Compute the sum of all ratio (ideally, it should be 1)
+local function gen_sum(self, i_s, i_e)
+    local sum, new_w = 0,0
+
+    for i = i_s or 1, i_e or #self.widgets do
+        if self._ratios[i] then
+            sum = sum + self._ratios[i]
+        else
+            new_w = new_w + 1
+        end
+    end
+
+    return sum, new_w
+end
+
+--- Make sure "diff" the sum is 1
+local function normalize(self)
+    local count = #self.widgets
+    if count == 0 then return end
+
+    -- Instead of adding "if" everywhere, just handle this common case
+    if count == 1 then
+        self._ratios = { 1 }
+        return
+    end
+
+    local sum, new_w = gen_sum(self)
+    local old_count  = #self.widgets - new_w
+
+    local to_add = (sum == 0) and 1 or (sum / old_count)
+
+    -- Make sure all widgets have a ratio
+    for i=1, #self.widgets do
+        if not self._ratios[i] then
+            self._ratios[i] = to_add
+        end
+    end
+
+    sum = sum + to_add*new_w
+
+    local delta, new_sum =  (1 - sum) / count,0
+
+    -- Increase or decrease each ratio so it the sum become 1
+    for i=1, #self.widgets do
+        self._ratios[i] = self._ratios[i] + delta
+        new_sum = new_sum + self._ratios[i]
+    end
+
+    -- Floating points summation isn't an exact science, worst on 32bit systems
+    assert(new_sum > 0.99 and new_sum < 1.01)
+end
+
+--- Layout a ratio layout. Each widget gets a share of the size proportional
+-- to its ratio
+-- @param context The context in which we are drawn.
+-- @param width The available width.
+-- @param height The available height.
+function ratio:layout(context, width, height)
+    local result = {}
+    local pos,spacing = 0, self._spacing
+    local num = #self.widgets
+    local total_spacing = (spacing*(num-1))
+
+    normalize(self)
+
+    for k, v in ipairs(self.widgets) do
+        local space = nil
+        local x, y, w, h
+
+        if self.dir == "y" then
+            space = height * self._ratios[k]
+            x, y = 0, util.round(pos)
+            w, h = width, floor(space)
+        else
+            space = width * self._ratios[k]
+            x, y = util.round(pos), 0
+            w, h = floor(space), height
+        end
+
+        table.insert(result, base.place_widget_at(v, x, y, w, h))
+
+        pos = pos + space + spacing
+
+        if (self.dir == "y" and pos-spacing >= height) or
+            (self.dir ~= "y" and pos-spacing >= width) then
```

Comment:
Why `pos-spacing >= height`? Shouldn't `util.round(pos) > height` be enough and more exact?